lib/Runtime/Provisioner/Provisioner.cpp

```diff
@@ -61,16 +61,35 @@ llvm::Error Provisioner::provision(DAGListTy &networks, Module &module,
                                    CompilationContext &cctx) {
   // Walk the networks and group by logicalDeviceId.
   std::map<DeviceIDTy, std::vector<DAGNode *>> logicalDevices;
+  // List of functions being added.
+  std::vector<std::string> localActiveNames;
   // For each network visit all the partitions (nodes) and add the node to each
   // logical device it is assigned to.
-  for (auto &network : networks) {
-    for (auto &node : network.nodes) {
-      for (auto logical : node->logicalDevices) {
-        auto it = logicalDevices.find(logical);
-        if (it != logicalDevices.end()) {
-          it->second.push_back(node.get());
-        } else {
-          logicalDevices.emplace(logical, std::vector<DAGNode *>{node.get()});
+  {
+    std::lock_guard<std::mutex> networkLock(functionsLock_);
```

Comment:
This makes me think the functionsLock_ protects `activeFunctions_` too since `functions_` is not touched in this block.

Reply:
It does, updated the comment to call this out.